lib/pbench/server/api/resources/query_apis/datasets/__init__.py

```diff
@@ -82,48 +80,39 @@ class IndexMapBase(ElasticBase):
     }
 
     def __init__(self, config: PbenchServerConfig, logger: Logger, *schemas: ApiSchema):
-        super().__init__(config, logger, *schemas)
-
         api_name = self.__class__.__name__
-
-        if len(schemas) != 1:
-            raise MissingDatasetNameParameter(
-                api_name, f"exactly one schema required: found {len(schemas)}"
-            )
-
-        self.schema = schemas[0]
-        dset = self.schema.get_param_by_type(
+        assert (
+            len(schemas) == 1
+        ), f"{api_name}: exactly one schema required: found {len(schemas)}"
+        dset = schemas[0].get_param_by_type(
```

Comment:
Yeah, I thought about that, and it's why this was an explicit exception rather than an `assert` originally. Generally, exceptions are important for production detection because they don't disappear with `-O`. On the other hand, this is a static condition we'd detect on the first PR commit that breaks it, where we presumably would never be running with `-O` and I don't expect it to ever fire in a production deployment. (Unless someone deploys `-O` without ever testing, in which case they kinda deserve what they get.)

So this should be just fine, and I know how much Pete loves asserts. 😀

Reply:
> this is a static condition we'd detect on the first PR commit that breaks it, where we presumably would never be running with -O and I don't expect it to ever fire in a production deployment.

Yeah.

> this should be just fine, and I know how much Pete loves asserts. 😀

Yup.

But, if we really never expect these to fire (e.g., because we have unit tests which check....), then perhaps we should just remove them from the product code and be done with the question....

Reply:
First, the existing code was using `MissingDatasetNameParameter()` which seemed odd to me to catch more than one schema used.

I am happy to use `assert` since this is what they are for, and that this code has nothing to do with the operation of the server but a check on the proper use of this class by a sub-class.

Since you seem to feel strongly about this, I'll just write these as "INTERNAL ERROR" if statements.

Reply:
For the record, although the change initially surprised me, I think it's ultimately a good idea... including the fact that it means we don't need to carry this check in production code. (Assuming we ever decide to run with `-O`.)

Reply:
Actually, I don't feel strongly, per se.  Unlike many cases, this probably is a reasonable use of `assert` in product code.  (I'm just against using `assert` in product code in general, because it so often ends up in poor results, so whenever I see one I'm inclined to question it; and, I'm resistant to adding new ones out of concern that it will pave the way for adding bad ones.)

However, if what we really want is a static (or static-like) check on the subclasses here, then that ought to be straightforward to implement that in a unit test, which would catch the problem without having to add even conditional checks to the product code.

(I also don't like the way the superclass initializer call got moved down and buried, but I understand why it happened; moving the asserts to a unit test would address that problem as well....)

Reply:
> FWIW, lines 24 & 25 are what prompted me to consider making this change.

Ha!  😆  OK, so Dave is contributing to the delinquency!  😀

Reply:
So I have a separate commit which uses `APIInternalError` raised instead of an assert.  It doesn't matter to me which one we go with.

Reply:
I dropped this back to using `assert`.